File: kmr.py
from __future__ import division
import numpy as np
from scipy.stats import binom
import matplotlib.pyplot as plt
import quantecon as qe


def kmr_markov_matrix_simultaneous(p, N, epsilon):
    P = np.empty((N+1, N+1))
    for n in range(N+1):
        P[n, :] = \
            (n/N < p) * binom.pmf(range(N+1), N, epsilon/2) + \
            (n/N == p) * binom.pmf(range(N+1), N, 1/2) + \
            (n/N > p) * binom.pmf(range(N+1), N, 1-epsilon/2)
    return P


def kmr_markov_matrix_sequential(p, N, epsilon):
    P = np.zeros((N+1, N+1))
    P[0, 0], P[0, 1] = 1 - epsilon * (1/2), epsilon * (1/2)
    for n in range(1, N):
        P[n, n-1] = (n/N) * (
            epsilon * (1/2) +
            (1 - epsilon) * (((n-1)/(N-1) < p) + ((n-1)/(N-1) == p) * (1/2))
        )
        P[n, n+1] = ((N-n)/N) * (
            epsilon * (1/2) +
            (1 - epsilon) * ((n/(N-1) > p) + (n/(N-1) == p) * (1/2))
        )
        P[n, n] = 1 - P[n, n-1] - P[n, n+1]
    P[N, N-1], P[N, N] = epsilon * (1/2), 1 - epsilon * (1/2)
    return P


class KMR2x2(object):

    def __init__(self, p, N, epsilon, revision='simultaneous'):

        self.p, self.N = p, N
        self._epsilon = epsilon
        self._mc = None

        if revision == 'simultaneous':
            self.kmr_markov_matrix = kmr_markov_matrix_simultaneous
        elif revision == 'sequential':
            self.kmr_markov_matrix = kmr_markov_matrix_sequential
        else:
            raise ValueError

    @property
    def epsilon(self):
        return self._epsilon

    @epsilon.setter
    def epsilon(self, value):
        self._epsilon = value
        self._mc = None

    @property
    def mc(self):
        if self._mc is None:
            P = self.kmr_markov_matrix(self.p, self.N, self.epsilon)
            self._mc = qe.MarkovChain(P)
        return self._mc

    def simulate(self, ts_length, init=None, num_reps=None):
        """
        Simulate the dynamics.

        Parameters
        ----------
        ts_length : scalar(int)
            Length of each simulation.

        init : scalar(int) or array_like(int, ndim=1),
               optional(default=None)
            Initial state(s). If None, the initial state is randomly
            drawn.

        num_reps : scalar(int), optional(default=None)
            Number of simulations. Relevant only when init is a scalar
            or None.

        Returns
        -------
        ndarray(int, ndim=1 or 2)
            Array containing the sample path(s), of shape (ts_length,)
            if init is a scalar (integer) or None and num_reps is None;
            of shape (k, ts_length) otherwise, where k = len(init) if
            init is an array_like, otherwise k = num_reps.

        """
        return self.mc.simulate(ts_length, init, num_reps)

    def plot_sample_path(self, x=None, ts_length=10**4, init=None,
                         ax=None, show=True):
        if x is None:
            x = self.simulate(ts_length, init)

        if show:
            fig, ax = plt.subplots()
        ax.plot(x, alpha=0.5)
        ax.set_ylim(0, self.N)
        ax.set_title(r'Sample path: $\varepsilon = {0}$'.format(self._epsilon))
        ax.set_xlabel('Time')
        ax.set_ylabel('State')
        if show:
            plt.show()

    def plot_emprical_dist(self, x=None, ts_length=10**4, init=None,
                           ax=None, show=True):
        if x is None:
            x = self.simulate(ts_length, init)

        if show:
            fig, ax = plt.subplots()
        hist, bins = np.histogram(x, self.N+1)
        ax.bar(range(self.N+1), hist, align='center')
        ax.set_title(r'Emprical distribution: $\varepsilon = {0}$'
                     .format(self.epsilon))
        ax.set_xlim(-0.5, self.N+0.5)
        ax.set_xlabel('State')
        ax.set_ylabel('Frequency')
        if show:
            plt.show()

    @property
    def stationary_dist(self):
        """
        Return an array containing the stationary distribution

        """
        return self.mc.stationary_distributions[0]

    def plot_stationary_dist(self, ax=None, show=True):
        if show:
            fig, ax = plt.subplots()
        ax.bar(range(self.N+1), self.stationary_dist, align='center')
        ax.set_xlim(-0.5, self.N+0.5)
        ax.set_ylim(0, 1)
        ax.set_title(r'Stationary distribution: $\varepsilon = {0}$'
                     .format(self.epsilon))
        ax.set_xlabel('State')
        ax.set_ylabel('Probability')
        if show:
            plt.show()

File: kmr.jl
using QuantEcon
using Distributions
using PyPlot
using StatsBase


function kmr_markov_matrix_simultaneous(p::Float64, N::Int, epsilon::Float64)
    binom_pdf_1 = pdf(Binomial(N, epsilon/2))
    binom_pdf_2 = pdf(Binomial(N, 1-epsilon/2))
    binom_pdf_tie = pdf(Binomial(N, 1/2))

    P = Array(Float64, N+1, N+1)
    for n in 0:N
        P[n+1, :] = (n/N < p) * binom_pdf_1 +
                    (n/N == p) * binom_pdf_tie +
                    (n/N > p) * binom_pdf_2
    end
    return P
end


function kmr_markov_matrix_sequential(p::Float64, N::Int, epsilon::Float64)
    P = zeros(N+1, N+1)
    P[1, 1], P[1, 2] = 1 - epsilon * (1/2), epsilon * (1/2)
    @inbounds for n in 1:N-1
        P[n+1, n] = (n/N) * (
            epsilon * (1/2) +
            (1 - epsilon) * (((n-1)/(N-1) < p) + ((n-1)/(N-1) == p) * (1/2))
        )
        P[n+1, n+2] = ((N-n)/N) * (
            epsilon * (1/2) +
            (1 - epsilon) * ((n/(N-1) > p) + (n/(N-1) == p) * (1/2))
        )
        P[n+1, n+1] = 1 - P[n+1, n] - P[n+1, n+2]
    end
    P[end, end-1], P[end, end] = epsilon * (1/2), 1 - epsilon * (1/2)
    return P
end


function kmr_markov_chain(p::Float64, N::Int, epsilon::Float64,
                          revision::Symbol)
    if revision == :simultaneous
        P = kmr_markov_matrix_simultaneous(p, N, epsilon)
    elseif revision == :sequential
        P = kmr_markov_matrix_sequential(p, N, epsilon)
    else
        throw(ArgumentError)
    end
    state_values = 0:N
    return MarkovChain(P, state_values)
end


type KMR2x2
    p::Float64
    N::Int
    epsilon::Float64
    revision::Symbol
    mc::MarkovChain{Float64,Matrix{Float64},UnitRange{Int}}
end

function KMR2x2(p::Float64, N::Int, epsilon::Float64;
                revision::Symbol=:simultaneous)
    mc = kmr_markov_chain(p, N, epsilon, revision)
    return KMR2x2(p, N, epsilon, revision, mc)
end

function epsilon!(kmr::KMR2x2, epsilon)
    kmr.epsilon = epsilon
    kmr.mc = kmr_markov_chain(kmr.p, kmr.N, epsilon, kmr.revision)
    return kmr
end


QuantEcon.simulate(kmr::KMR2x2, ts_length::Int; init::Int=rand(0:kmr.N)) =
    simulate(kmr.mc, ts_length, init=init+1)

function simulate_cross_section(kmr::KMR2x2, ts_length::Int;
                                init::Vector{Int}=rand(0:kmr.N, num_reps),
                                num_reps::Int=1)
    X = Array(Int, ts_length, num_reps)
    simulate!(X, kmr.mc, init=init.+1)
    return X
end

function plot_sample_path(kmr::KMR2x2, x::Vector{Int};
                          ax=Union{}, show_plot=true)
    if show_plot
        fig, ax = subplots()
    end
    ax[:plot](x, alpha=0.5)
    ax[:set_ylim](0, kmr.N)
    ax[:set_title]("Sample path: \$\\varepsilon = $(kmr.epsilon)\$")
    ax[:set_xlabel]("Time")
    ax[:set_ylabel]("State")
    plt[:show]()
end

function plot_sample_path(kmr::KMR2x2, ts_length::Int; init::Int=rand(0:kmr.N),
                          ax=Union{}, show_plot=true)
    x = simulate(kmr, ts_length, init=init)
    plot_sample_path(kmr, x, ax=ax, show_plot=show_plot)
end


function plot_empirical_dist(kmr::KMR2x2, x::Vector{Int};
                             ax=Union{}, show_plot=true)
    h = fit(Histogram, x, 0:kmr.N+1, closed=:left)
    if show_plot
        fig, ax = subplots()
    end
    ax[:bar](0:kmr.N, h.weights, align="center")
    ax[:set_title]("Empirical distribution: \$\\varepsilon = $(kmr.epsilon)\$")
    ax[:set_xlim](-0.5, kmr.N+0.5)
    ax[:set_xlabel]("State")
    ax[:set_ylabel]("Frequency")
    plt[:show]()
end

function plot_empirical_dist(kmr::KMR2x2, ts_length::Int;
                             init::Int=rand(0:kmr.N),
                             ax=Union{}, show_plot=true)
    x = simulate(kmr, ts_length, init=init)
    plot_empirical_dist(kmr, x, ax=ax, show_plot=show_plot)
end


stationary_dist(kmr::KMR2x2) = stationary_distributions(kmr.mc)[1]

function plot_stationary_dist(kmr::KMR2x2; ax=Union{}, show_plot=true)
    if show_plot
        fig, ax = subplots()
    end
    ax[:bar](0:kmr.N, stationary_dist(kmr), align="center")
    ax[:set_title]("Stationary distribution: \$\\varepsilon = $(kmr.epsilon)\$")
    ax[:set_xlim](-0.5, kmr.N+0.5)
    ax[:set_xlabel]("State")
    ax[:set_ylabel]("Probability")
    plt[:show]()
end
